buckets-mdapi/src/object/get.rs

```diff
@@ -64,33 +70,51 @@ pub(crate) fn action(
 fn do_get(
     method: &str,
     payload: &GetObjectPayload,
-    mut conn: &mut PostgresConnection,
+    conn: &mut PostgresConnection,
     metrics: &RegisteredMetrics,
     log: &Logger,
-) -> Result<Option<ObjectResponse>, String> {
-    let sql = get_sql(payload.vnode);
+) -> Result<Option<ObjectResponse>, BucketsMdapiError> {
+    let mut txn = (*conn).transaction().map_err(|e| {
+        BucketsMdapiError::PostgresError(e.to_string())
+    })?;
+
+    let get_sql = sql::get_sql(payload.vnode);
 
-    sql::query(
-        sql::Method::ObjectGet,
-        &mut conn,
-        sql.as_str(),
+    conditional::request(
+        &mut txn,
         &[&payload.owner, &payload.bucket_id, &payload.name],
+        payload.vnode,
+        &payload.conditions,
         metrics,
         log,
     )
-    .map_err(|e| e.to_string())
-    .and_then(|rows| response(method, &rows))
-}
+    .and_then(|rows| {
+        /*
+         * conditional::request will perform the same query as this method, so if it returns
+         * Some(rows) then it means the request was both conditional and a success, so we can just
+         * return those rows as-is.
+         */
+        if let Some(r) = rows {
+            return Ok(r);
```

Comment:
Ah, didn't plan on that. I thought the rollback happened only when `txn` gets dropped at the end of `do_get` if `txt.commit()` hasn't been called, and that my early return would still lead to that commit happening. Maybe I've misunderstood this a bit. Mind pointing out where the rollback happens?

You're right though on the whole transaction handling thing. Your initial recommendation is probably the way to go. My main problem with having `go_get` check the conditions itself was that it would also have to check the row count for 0 or >1 rows, but maybe if I have `Conditions::check()` take `RowSlice` instead I can have that method do that row count check also. I'll have a little play around!

Reply:
Ok, I've taken another go at this in https://github.com/chudley/manta-buckets-mdapi/commit/17b97720701cca759115a7517c56a4dc2220d5f6. `do_get` now just runs the SQL for a get and passes an `ObjectResponse` to the conditional check, which has also been changed to accept that instead.

Reply:
Yeah, I like the way that looks now. Very nice!